File: 123-best-time-to-buy-and-sell-stock-iii/123-best-time-to-buy-and-sell-stock-iii.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        dp_1_hold, dp_1_not_hold = float('-inf'), 0
        dp_2_hold, dp_2_not_hold = float('-inf'), 0
        
        for stock_price in prices:    
            dp_1_hold = max(dp_1_hold, -stock_price)
            dp_1_not_hold = max(dp_1_not_hold, dp_1_hold + stock_price)
            
            dp_2_hold = max(dp_2_hold, dp_1_not_hold - stock_price)
            dp_2_not_hold = max(dp_2_not_hold, dp_2_hold + stock_price)
        
        return dp_2_not_hold

File: 123-best-time-to-buy-and-sell-stock-iii/123-best-time-to-buy-and-sell-stock-iii.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int bestBuy1 = INT_MIN;
        int bestSell1 = 0;
        int bestBuy2 = INT_MIN;
        int bestSell2 = 0;
        
        for(auto p: prices){
            bestBuy1 = max(-p, bestBuy1);
            bestSell1 = max(bestSell1, bestBuy1 + p);
            
            bestBuy2 = max(bestSell1 - p, bestBuy2);
            bestSell2 = max(bestSell2, bestBuy2 + p);
        }
        
        return bestSell2;
    }
};